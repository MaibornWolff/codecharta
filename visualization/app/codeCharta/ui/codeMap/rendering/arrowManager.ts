import * as THREE from "three";
import {node} from "./node";
import {renderSettings} from "./renderSettings";
import {Group, Object3D, Vector3} from "three";
import {CodeMapDependency} from "../../../core/data/model/CodeMap";

export class ArrowManager {
    private parentObjectInScene: THREE.Object3D;
    private arrows: THREE.Object3D[];

    constructor(argParentObjectInScene: THREE.Object3D) {
        this.parentObjectInScene = argParentObjectInScene;
        this.arrows = new Array<THREE.Object3D>();
    }

    private makeArrowFromBezier(bezier: THREE.CubicBezierCurve3,
                                hex: number = 0,
                                headLength: number = 10,
                                headWidth: number = 10,
                                bezierPoints: number = 50): THREE.Object3D {

        let points = bezier.getPoints(bezierPoints);

        // arrowhead
        let dir = points[points.length - 1].clone().sub(points[points.length - 2].clone());
        dir.normalize();
        let origin = points[points.length - 1].clone();
        let arrowHelper = new THREE.ArrowHelper(dir, origin, 0, hex, headLength, headWidth);

        // curve
        let geometry = new THREE.BufferGeometry();
        geometry.setFromPoints(points);
        let material = new THREE.LineBasicMaterial({color: hex, linewidth: 1});
        let curveObject = new THREE.Line(geometry, material);

        //combine
        curveObject.add(arrowHelper);
        return curveObject;
    }

    public clearArrows() {
        this.arrows = [];
        while (this.parentObjectInScene.children.length > 0) {
            this.parentObjectInScene.children.pop();
        }
    }

    public addCodeMapDependenciesFromOriginAsArrows(origin: node, nodes: node[], deps: CodeMapDependency[], settings: renderSettings) {
        let resDeps: CodeMapDependency[] = [];
        let originPath = this.getPathFromNode(origin);
        for(let i=0; i<deps.length; i++){
            if(deps[i].node === originPath) {
                resDeps.push(deps[i]);
            }
        }
        this.addCodeMapDependenciesAsArrows(nodes, resDeps, settings);
    }

    public addCodeMapDependenciesAsArrows(nodes: node[], deps: CodeMapDependency[], settings: renderSettings) {

        let map = this.getPathToNodeMap(nodes);

        for (let i = 0; i < deps.length; i++) {
            let originNode: node = map.get(deps[i].node);
            let targetNode: node = map.get(deps[i].dependsOn);
            if (originNode && targetNode) {
                this.addArrow(targetNode, originNode, settings);
            }
        }

    }

    private getPathToNodeMap(nodes: node[]): Map<string, node> {
        let map = new Map<string, node>();

        for (let node of nodes) {
            map.set(this.getPathFromNode(node), node);
        }

        return map;
    }

    private getPathFromNode(node: node): string {
        let current: node = node;
        let path = "";
        while (current) {
            path = "/" + current.name + path;
            current = current.parent;
        }
        return path;
    }

    public addArrow(arrowTargetNode: node, arrowOriginNode: node, settings: renderSettings): void {

        if (arrowTargetNode.attributes && arrowTargetNode.attributes[settings.heightKey] && arrowOriginNode.attributes && arrowOriginNode.attributes[settings.heightKey]) {

            let xTarget: number = arrowTargetNode.x0 - settings.mapSize * 0.5;
            let yTarget: number = arrowTargetNode.z0;
            let zTarget: number = arrowTargetNode.y0 - settings.mapSize * 0.5;

            let wTarget: number = arrowTargetNode.width;
            let hTarget: number = arrowTargetNode.height;
            let lTarget: number = arrowTargetNode.length;

            let xOrigin: number = arrowOriginNode.x0 - settings.mapSize * 0.5;
            let yOrigin: number = arrowOriginNode.z0;
            let zOrigin: number = arrowOriginNode.y0 - settings.mapSize * 0.5;

            let wOrigin: number = arrowOriginNode.width;
            let hOrigin: number = arrowOriginNode.height;
            let lOrigin: number = arrowOriginNode.length;

            let curve = new THREE.CubicBezierCurve3(
                new THREE.Vector3(xOrigin + wOrigin / 2, yOrigin + hOrigin, zOrigin + lOrigin / 2),
                new THREE.Vector3(xOrigin + wOrigin / 2, Math.max(yOrigin + hOrigin, yTarget + hTarget) + settings.mapSize, zOrigin + lOrigin / 2),
                new THREE.Vector3(xTarget + wTarget / 2, Math.max(yOrigin + hOrigin, yTarget + hTarget) + settings.mapSize, zTarget + lTarget / 2),
                new THREE.Vector3(xTarget + wTarget / 2, yTarget + hTarget, zTarget + lTarget / 2)
            );

            let arrow: THREE.Object3D = this.makeArrowFromBezier(curve);

            this.parentObjectInScene.add(arrow);
            this.arrows.push(arrow);

        }

    }

    public scale(x: number, y: number, z: number) {
        for (let arrow of this.arrows) {
            arrow.scale.x = x;
            arrow.scale.y = y;
            arrow.scale.z = z;
        }
    }

}