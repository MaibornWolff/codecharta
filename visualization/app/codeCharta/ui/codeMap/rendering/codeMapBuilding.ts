import * as THREE from "three";
import {MapColors} from "./renderSettings";
import {node} from "./node";

export class codeMapBuilding {
    public id : number;
    public boundingBox : THREE.Box3;
    public color : number;
    public node : node;
    public parent?: codeMapBuilding;
    public name?: string;

    constructor(id : number, box : THREE.Box3, node : node, color? : number)
    {
        this.id = id;
        this.boundingBox = box;
        this.color = color ? color : MapColors.defaultC;
        this.node = node;
    }
}