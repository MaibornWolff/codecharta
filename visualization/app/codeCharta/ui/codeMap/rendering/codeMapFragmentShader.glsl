varying vec3 vColor;
varying vec3 vDeltaColor;
varying vec3 worldNormal;
varying vec3 vLightFront;
varying highp float vDelta;
varying vec2 vUV;

uniform vec3 ambientLightColor;

#define RECIPROCAL_PI 0.31830988618
#define PI 3.14159265359

struct ReflectedLight {
    vec3 directDiffuse;
    vec3 directSpecular;
    vec3 indirectDiffuse;
    vec3 indirectSpecular;
};

vec3 getAmbientLightIrradiance(const in vec3 ambientLightColor) {
    vec3 irradiance = ambientLightColor;
    irradiance *= PI;
    return irradiance;
}

vec3 BRDF_Diffuse_Lambert (const in vec3 diffuseColor) {
    return RECIPROCAL_PI * diffuseColor;
}

bool normalPointingUp(const in vec3 normal)
{
    return normal.y > 0.9;
}

struct Rect {
    vec2 min;
    vec2 max;
};

bool uvCoordInRectangle(const in vec2 uv, const in Rect r)
{
    return uv.x >= r.min.x && uv.x < r.max.x && uv.y >= r.min.y && uv.y < r.max.y;
}

void main() {
    vec4 diffuseColor = vec4(vColor, 1.0);

    const float epsilon = 0.5;
    const float minDelta = 0.001;

    bool isTop = normalPointingUp(worldNormal);

    if (abs(vDelta) > minDelta && (vUV.y > 1.0 - abs(vDelta) || isTop))
    {
        diffuseColor.xyz = vDeltaColor;
    }

    ReflectedLight reflectedLight = ReflectedLight(vec3(0.0), vec3(0.0), vec3(0.0), vec3(0.0));
    reflectedLight.indirectDiffuse = getAmbientLightIrradiance(ambientLightColor);
    reflectedLight.indirectDiffuse *= BRDF_Diffuse_Lambert(diffuseColor.rgb);

    reflectedLight.directDiffuse = vLightFront;
    reflectedLight.directDiffuse *= BRDF_Diffuse_Lambert(diffuseColor.rgb);

    vec3 outgoingLight = reflectedLight.directDiffuse + reflectedLight.indirectDiffuse;

    gl_FragColor = vec4(outgoingLight, diffuseColor.a);
}
