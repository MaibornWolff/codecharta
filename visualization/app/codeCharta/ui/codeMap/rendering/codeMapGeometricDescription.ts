import {codeMapBuilding} from "./codeMapBuilding";
import * as THREE from "three";

export interface intersectionResult {
    intersectionFound: boolean;
    building?: codeMapBuilding;
}

export class codeMapGeometricDescription {
    private _buildings: codeMapBuilding[];
    private mapSize: number;
    private scales: THREE.Vector3;

    constructor(mapSize: number) {
        this._buildings = new Array<codeMapBuilding>();
        this.mapSize = mapSize;
        this.scales = new THREE.Vector3(1, 1, 1);
    }

    get buildings(): codeMapBuilding[] {
        return this._buildings;
    }

    add(building: codeMapBuilding): void {
        this._buildings.push(building);
    }

    setScales(scales: THREE.Vector3) {
        this.scales = scales;
    }

    private rayIntersectsAxisAlignedBoundingBox(ray: THREE.Ray, box: THREE.Box3): boolean {
        let tx1 = (box.min.x - ray.origin.x) * (1 / ray.direction.x);
        let tx2 = (box.max.x - ray.origin.x) * (1 / ray.direction.x);

        let tmin = Math.min(tx1, tx2);
        let tmax = Math.max(tx1, tx2);

        let ty1 = (box.min.y - ray.origin.y) * (1 / ray.direction.y);
        let ty2 = (box.max.y - ray.origin.y) * (1 / ray.direction.y);

        tmin = Math.max(tmin, Math.min(ty1, ty2));
        tmax = Math.min(tmax, Math.max(ty1, ty2));

        return tmax >= tmin;
    }

    intersect(ray: THREE.Ray): intersectionResult {

        let intersectedBuilding: codeMapBuilding | null = null;
        let leastIntersectedDistance: number = Infinity;

        let boxTranslation = new THREE.Vector3(-this.mapSize * this.scales.x * 0.5, 0.0, -this.mapSize * this.scales.z * 0.5);

        for (let building of this._buildings) {

            //Pre Transformation
            let box: THREE.Box3 = building.boundingBox.clone();

            box.min.x *= this.scales.x;
            box.min.y *= this.scales.y;
            box.min.z *= this.scales.z;

            box.max.x *= this.scales.x;
            box.max.y *= this.scales.y;
            box.max.z *= this.scales.z;

            box.translate(boxTranslation);

            if (this.rayIntersectsAxisAlignedBoundingBox(ray, box)) {

                let intersectionPoint: THREE.Vector3 = ray.intersectBox(box);

                if (intersectionPoint) {
                    let intersectionDistance: number = intersectionPoint.distanceTo(ray.origin);

                    if (intersectionDistance < leastIntersectedDistance) {
                        leastIntersectedDistance = intersectionDistance;
                        intersectedBuilding = building;
                    }
                }

            }

        }

        if (intersectedBuilding) {
            return {
                intersectionFound: true,
                building: intersectedBuilding
            };
        }
        else {
            return {
                intersectionFound: false
            };
        }
    }
}