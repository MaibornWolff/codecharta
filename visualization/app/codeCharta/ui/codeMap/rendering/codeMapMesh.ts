import * as THREE from "three";

import {codeMapShaderStrings} from "./codeMapShaderStrings";
import {geometryGenerator, buildResult} from "./geometryGenerator";
import {codeMapGeometricDescription} from "./codeMapGeometricDescription";
import {intersectionResult} from "./codeMapGeometricDescription";
import {codeMapBuilding} from "./codeMapBuilding";
import {renderingUtil} from "./renderingUtil";
import {node} from "./node";

import {MapColors} from "./renderSettings";
import {renderSettings} from "./renderSettings";

interface threeUniform {
    type : string;
    value : any;
}

interface codeMapLightingParams {
    numHighlights : threeUniform;
    numSelections : threeUniform;
    highlightColor : threeUniform;
    highlightedIndices : threeUniform;
    selectedColor : threeUniform;
    selectedIndices : threeUniform;
    emissive : threeUniform;
    deltaColorPositive : threeUniform;
    deltaColorNegative : threeUniform;
}

export interface mousePos {
    x : number;
    y : number;
}

export class CodeMapMesh {
    private threeMesh : THREE.Mesh;
    private material : THREE.ShaderMaterial;
    private geomGen : geometryGenerator;
    private mapGeomDesc : codeMapGeometricDescription;

    private nodes : node[];

    private currentlyHighlighted : codeMapBuilding[] | null;
    private currentlySelected : codeMapBuilding[] | null;

    public settings : renderSettings;

    private lightingParams : codeMapLightingParams = {
        numHighlights : {type : "f", value : 0.0},
        highlightColor : {type : "v3", value : renderingUtil.colorToVec3(0x666666)},
        highlightedIndices : {type : "fv1", value : []},

        numSelections : {type : "f", value : 0.0},
        selectedColor : {type : "f", value : renderingUtil.colorToVec3(MapColors.selected)},
        selectedIndices : {type : "fv1", value : []},

        deltaColorPositive : {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)},
        deltaColorNegative : {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)},

        emissive : {type : "v3", value : new THREE.Vector3(0.0, 0.0, 0.0)}
    };

    constructor(nodes : node[], settings : renderSettings)
    {
        this.nodes = nodes;

        this.initMaterial(settings);

        this.geomGen = new geometryGenerator();
        let buildRes : buildResult = this.geomGen.build(this.nodes, this.material, settings);

        this.threeMesh = buildRes.mesh;
        this.mapGeomDesc = buildRes.desc;

        this.settings = settings;
    }

    private initMaterial(settings : renderSettings) : void
    {

        if(settings.deltaColorFlipped) {
            this.setDeltaColorsFlipped();
        } else {
            this.setDeltaColorsUnflipped();
        }

        let uniforms = THREE.UniformsUtils.merge([THREE.UniformsLib["lights"], this.lightingParams]);

        let shaderCode : codeMapShaderStrings = new codeMapShaderStrings();

        this.material =  new THREE.ShaderMaterial({
            vertexShader: shaderCode.vertexShaderCode,
            fragmentShader: shaderCode.fragmentShaderCode,
            lights : true,
            uniforms : uniforms
        });
    }

    private setDeltaColorsFlipped() {
        this.lightingParams.deltaColorPositive = {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)};
        this.lightingParams.deltaColorNegative = {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)};
    }


    private setDeltaColorsUnflipped() {
        this.lightingParams.deltaColorPositive = {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)};
        this.lightingParams.deltaColorNegative = {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)};
    }

    public getThreeMesh() : THREE.Mesh {
        return this.threeMesh;
    }

    public setHighlighted(buildings : codeMapBuilding[], color? : number)
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.highlightedIndices.value = buildings.map((b : codeMapBuilding) => {return b.id;});
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numHighlights.value = buildings.length;

        if (color) {
            this.lightingParams.highlightColor.value = renderingUtil.colorToVec3(color);
        }

        this.currentlyHighlighted = buildings;
    }

    public setSelected(buildings : codeMapBuilding[], color? : number)
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.selectedIndices.value = buildings.map((b : codeMapBuilding) => {return b.id;});
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numSelections.value  = buildings.length;
        
        if (color) {
            this.lightingParams.selectedColor.value = renderingUtil.colorToVec3(color);
        }

        this.currentlySelected = buildings;
    }

    public getCurrentlyHighlighted() : codeMapBuilding[] | null
    {
        return this.currentlyHighlighted;
    }

    public getCurrentlySelected() : codeMapBuilding[] | null
    {
        return this.currentlySelected;
    }

    public clearHighlight()
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numHighlights.value = 0.0;
        this.currentlyHighlighted = null;
    }

    public clearSelected()
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numSelections.value = 0.0;
        this.currentlySelected = null;
    }

    public getMeshDescription() : codeMapGeometricDescription
    {
        return this.mapGeomDesc;
    }

    private calculatePickingRay(mouse : mousePos, camera : THREE.Camera) : THREE.Ray
    {

        let ray : THREE.Ray = new THREE.Ray();
        ray.origin.setFromMatrixPosition(camera.matrixWorld);
        ray.direction.set(mouse.x, mouse.y, 0.5).unproject(camera).sub(ray.origin).normalize();

        return ray;
    }

    public checkMouseRayMeshIntersection(mouse : mousePos, camera : THREE.Camera) : intersectionResult
    {
        let ray : THREE.Ray = this.calculatePickingRay(mouse, camera);
        return this.getMeshDescription().intersect(ray);
    }

    public setScale(x : number, y : number, z : number)
    {
        this.mapGeomDesc.setScales(new THREE.Vector3(x, y, z));
    }
}