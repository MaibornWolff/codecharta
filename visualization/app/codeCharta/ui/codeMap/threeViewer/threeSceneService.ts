import { AmbientLight, DirectionalLight, Scene, Group } from "three"
import { CodeMapMesh } from "../rendering/codeMapMesh"
import { CodeMapBuilding } from "../rendering/codeMapBuilding"
import { CodeMapPreRenderServiceSubscriber, CodeMapPreRenderService } from "../codeMap.preRender.service"
import { IRootScopeService } from "angular"
import { StoreService } from "../../../state/store.service"

export interface BuildingSelectedEventSubscriber {
	onBuildingSelected(selectedBuilding?: CodeMapBuilding)
}

export interface BuildingDeselectedEventSubscriber {
	onBuildingDeselected()
}

export interface CodeMapMeshChangedSubscriber {
	onCodeMapMeshChanged(mapMesh: CodeMapMesh)
}

export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
	private static readonly BUILDING_SELECTED_EVENT = "building-selected"
	private static readonly BUILDING_DESELECTED_EVENT = "building-deselected"
	private static readonly CODE_MAP_MESH_CHANGED_EVENT = "code-map-mesh-changed"

	scene: Scene
	labels: Group
	edgeArrows: Group
	mapGeometry: Group
	private readonly lights: Group
	private mapMesh: CodeMapMesh

	private selected: CodeMapBuilding = null
	private highlighted: CodeMapBuilding[] = []

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		CodeMapPreRenderService.subscribe(this.$rootScope, this)

		this.scene = new Scene()
		this.mapGeometry = new Group()
		this.lights = new Group()
		this.labels = new Group()
		this.edgeArrows = new Group()

		this.initLights()

		this.scene.add(this.mapGeometry)
		this.scene.add(this.edgeArrows)
		this.scene.add(this.labels)
		this.scene.add(this.lights)
	}

	onRenderMapChanged() {
		this.reselectBuilding()
	}

	highlightBuildings() {
		const state = this.storeService.getState()
		this.getMapMesh().highlightBuilding(this.highlighted, this.selected, state)
	}

	highlightSingleBuilding(building: CodeMapBuilding) {
		this.highlighted = []
		this.addBuildingToHighlightingList(building)
		this.highlightBuildings()
	}

	addBuildingToHighlightingList(building: CodeMapBuilding) {
		this.highlighted.push(building)
	}

	clearHighlight() {
		if (this.getMapMesh()) {
			this.getMapMesh().clearHighlight(this.selected)
			this.highlighted = []
		}
	}

	selectBuilding(building: CodeMapBuilding) {
		const color = this.storeService.getState().appSettings.mapColors.selected
		this.getMapMesh().selectBuilding(building, color)
		this.selected = building
		this.highlightBuildings()
		this.$rootScope.$broadcast(ThreeSceneService.BUILDING_SELECTED_EVENT, this.selected)
	}

	clearSelection() {
		if (this.selected) {
			this.getMapMesh().clearSelection(this.selected)
			this.$rootScope.$broadcast(ThreeSceneService.BUILDING_DESELECTED_EVENT)
		}
		if (this.highlighted.length > 0) {
			this.highlightBuildings()
		}
		this.selected = null
	}

	initLights() {
		const ambilight = new AmbientLight(0x707070) // soft white light
		const light1 = new DirectionalLight(0xe0e0e0, 1)
		light1.position.set(50, 10, 8).normalize()

		const light2 = new DirectionalLight(0xe0e0e0, 1)
		light2.position.set(-50, 10, -8).normalize()

		this.lights.add(ambilight)
		this.lights.add(light1)
		this.lights.add(light2)
	}

	setMapMesh(mesh: CodeMapMesh) {
		const { mapSize } = this.storeService.getState().treeMap
		this.mapMesh = mesh

		while (this.mapGeometry.children.length > 0) {
			this.mapGeometry.remove(this.mapGeometry.children[0])
		}

		this.mapGeometry.position.x = -mapSize
		this.mapGeometry.position.y = 0
		this.mapGeometry.position.z = -mapSize

		this.mapGeometry.add(this.mapMesh.getThreeMesh())
		this.notifyMapMeshChanged()
	}

	getMapMesh() {
		return this.mapMesh
	}

	scale() {
		const { mapSize } = this.storeService.getState().treeMap
		const scale = this.storeService.getState().appSettings.scaling

		this.mapGeometry.scale.set(scale.x, scale.y, scale.z)
		this.mapGeometry.position.set(-mapSize * scale.x, 0, -mapSize * scale.z)
		this.mapMesh.setScale(scale)
	}

	getSelectedBuilding() {
		return this.selected
	}

	getHighlightedBuilding() {
		return this.highlighted[0]
	}

	getHighlightedNode() {
		if (this.getHighlightedBuilding()) {
			return this.getHighlightedBuilding().node
		}
		return null
	}

	private reselectBuilding() {
		if (this.selected) {
			const buildingToSelect: CodeMapBuilding = this.getMapMesh().getBuildingByPath(this.selected.node.path)
			if (buildingToSelect) {
				this.selectBuilding(buildingToSelect)
			}
		}
	}

	private notifyMapMeshChanged() {
		this.$rootScope.$broadcast(ThreeSceneService.CODE_MAP_MESH_CHANGED_EVENT, this.mapMesh)
	}

	static subscribeToBuildingDeselectedEvents($rootScope: IRootScopeService, subscriber: BuildingDeselectedEventSubscriber) {
		$rootScope.$on(this.BUILDING_DESELECTED_EVENT, () => {
			subscriber.onBuildingDeselected()
		})
	}

	static subscribeToBuildingSelectedEvents($rootScope: IRootScopeService, subscriber: BuildingSelectedEventSubscriber) {
		$rootScope.$on(this.BUILDING_SELECTED_EVENT, (_event, selectedBuilding: CodeMapBuilding) => {
			subscriber.onBuildingSelected(selectedBuilding)
		})
	}

	static subscribeToCodeMapMeshChangedEvent($rootScope: IRootScopeService, subscriber: CodeMapMeshChangedSubscriber) {
		$rootScope.$on(this.CODE_MAP_MESH_CHANGED_EVENT, (_event, mapMesh: CodeMapMesh) => {
			subscriber.onCodeMapMeshChanged(mapMesh)
		})
	}
}
