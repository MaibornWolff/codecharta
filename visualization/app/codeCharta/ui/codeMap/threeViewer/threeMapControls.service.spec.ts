import { TestBed } from "@angular/core/testing"
import { StoreModule } from "@ngrx/store"
import { ThreeMapControlsService } from "./threeMapControls.service"
import { ThreeCameraService } from "./threeCamera.service"
import { ThreeSceneService } from "./threeSceneService"
import { Box3, BoxGeometry, Group, Mesh, MOUSE, PerspectiveCamera, Sphere, Vector3 } from "three"
import { ThreeRendererService } from "./threeRenderer.service"
import { wait } from "../../../util/testUtils/wait"
import { appReducers, setStateMiddleware } from "../../../state/store/state.manager"
import { MapControls } from "three/examples/jsm/controls/MapControls"
import { take } from "rxjs"
import { CodeMapMesh } from "../rendering/codeMapMesh"
import { fireEvent } from "@testing-library/angular"

describe("ThreeMapControlsService", () => {
    let threeMapControlsService: ThreeMapControlsService
    let threeCameraService: ThreeCameraService
    let threeSceneService: ThreeSceneService
    let threeRendererService: ThreeRendererService

    let vector: Vector3

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [StoreModule.forRoot(appReducers, { metaReducers: [setStateMiddleware] })]
        })
        restartSystem()
        rebuildService()
        withMockedThreeCameraService()
        withMockedThreeSceneService()
        withMockedControlService()
    })

    function restartSystem() {
        threeCameraService = TestBed.inject(ThreeCameraService)
        threeSceneService = TestBed.inject(ThreeSceneService)
        threeRendererService = TestBed.inject(ThreeRendererService)

        vector = new Vector3(5.711_079_128_159_569, 5.711_079_128_159_569, 0)
    }

    function withMockedThreeCameraService() {
        const camera = new PerspectiveCamera(100, 0, 0, 0)
        camera.position.set(vector.x, vector.y, vector.z)
        threeCameraService.camera = camera
    }

    function withMockedThreeSceneService() {
        threeSceneService.scene.add = jest.fn()
        threeSceneService.scene.remove = jest.fn()
        threeSceneService.mapGeometry = new Group().add(new Mesh(new BoxGeometry(10, 10, 10)))
    }

    function withMockedControlService() {
        threeMapControlsService.controls = {
            target: new Vector3(1, 1, 1),
            minDistance: 100,
            maxDistance: 1_000_000
        } as unknown as MapControls
        threeMapControlsService.controls.update = jest.fn()
        threeMapControlsService.controls.addEventListener = jest.fn()
    }

    function rebuildService() {
        threeMapControlsService = new ThreeMapControlsService(threeCameraService, threeSceneService, threeRendererService)
    }

    describe("init", () => {
        it("should initialize MapControls and set up event listeners correctly", () => {
            const domElement = document.createElement("canvas")
            withMockedThreeCameraService()
            const addEventListenerMock = jest.fn()
            jest.spyOn(MapControls.prototype, "addEventListener").mockImplementation(addEventListenerMock)

            threeMapControlsService.controls = new MapControls(threeCameraService.camera, domElement)

            threeMapControlsService.init(domElement)

            expect(threeMapControlsService.controls).toBeDefined()
            expect(threeMapControlsService.controls.mouseButtons.LEFT).toBe(MOUSE.ROTATE)
            expect(threeMapControlsService.controls.mouseButtons.MIDDLE).toBe(MOUSE.DOLLY)
            expect(threeMapControlsService.controls.mouseButtons.RIGHT).toBe(MOUSE.PAN)
            expect(addEventListenerMock).toHaveBeenCalledWith("change", expect.any(Function))
        })
    })

    it("rotateCameraInVectorDirection ", () => {
        threeMapControlsService.controls = {
            target: new Vector3(0, 0, 0)
        } as unknown as MapControls
        const vector = { x: 0, y: 1, z: 0 }

        threeMapControlsService.rotateCameraInVectorDirection(vector.x, vector.y, vector.z)

        expect(threeSceneService.scene.add).toHaveBeenCalled()
        expect(threeSceneService.scene.remove).toHaveBeenCalled()

        expect(threeCameraService.camera.position).toMatchSnapshot()
    })

    it("should change zoom to follow cursor when zooming in", () => {
        const domElement = document.createElement("canvas")
        threeMapControlsService.controls = new MapControls(threeCameraService.camera, domElement)
        threeMapControlsService.init(domElement)

        fireEvent.wheel(window, { deltaY: -2 })

        expect(threeMapControlsService.controls.zoomToCursor).toBeTruthy()
    })

    it("should change zoom to default setting when zooming out again", () => {
        const domElement = document.createElement("canvas")
        threeMapControlsService.controls = new MapControls(threeCameraService.camera, domElement)
        threeMapControlsService.init(domElement)
        threeMapControlsService.controls.zoomToCursor = true

        fireEvent.wheel(window, { deltaY: 2 })

        expect(threeMapControlsService.controls.zoomToCursor).toBeFalsy()
    })

    describe("setControlTarget", () => {
        it("should set the controlTarget to the store cameraTarget", () => {
            const result: Vector3 = new Vector3(1, 1, 1)
            threeMapControlsService.setControlTarget(new Vector3(1, 1, 1))

            expect(threeMapControlsService.controls.target).toEqual(result)
        })
    })

    describe("autoFitTo", () => {
        it("should auto fit map to its origin value ", async () => {
            threeCameraService.camera.position.set(0, 0, 0)

            threeMapControlsService.autoFitTo()
            await wait(0)

            expect(threeCameraService.camera.position).toEqual(new Vector3(5.423_365_845_416_211, 5.423_365_845_416_211, 0))
        })

        it("should call an control update", async () => {
            threeCameraService.camera.lookAt = jest.fn()

            threeMapControlsService.autoFitTo()
            await wait(0)

            expect(threeCameraService.camera.lookAt).toHaveBeenCalledWith(threeMapControlsService.controls.target)
        })

        it("should auto fit map to its original value ", async () => {
            threeCameraService.camera.updateProjectionMatrix = jest.fn()

            threeMapControlsService.autoFitTo()
            await wait(0)

            expect(threeMapControlsService.controls.update).toHaveBeenCalled()
            expect(threeCameraService.camera.updateProjectionMatrix).toHaveBeenCalled()
        })

        it("should return early if boundingSphere.radius is -1", async () => {
            jest.spyOn(threeMapControlsService, "getBoundingSphere").mockReturnValue(new Sphere())
            threeCameraService.camera.position.set(0, 0, 0)

            threeMapControlsService.autoFitTo()
            await wait(0)

            expect(threeCameraService.camera.position).toEqual(new Vector3(0, 0, 0))
        })
    })

    describe("setZoomPercentage", () => {
        beforeEach(() => {
            rebuildService()
            withMockedControlService()
        })

        it("should set the correct zoom percentage and emit it", async () => {
            const initialZoom = 50

            threeMapControlsService.setZoomPercentage(initialZoom)

            let actualZoomPercentage: number
            threeMapControlsService.zoomPercentage$.pipe(take(1)).subscribe(zoomPercentage => (actualZoomPercentage = zoomPercentage))

            expect(actualZoomPercentage).toBe(initialZoom)
        })
    })

    describe("updateZoomPercentage", () => {
        it("should update the zoom percentage correctly", () => {
            threeCameraService.camera.position.set(0, 0, 500_000)
            threeMapControlsService.updateZoomPercentage()

            const distance = threeCameraService.camera.position.length()
            const expectedZoomPercentage = threeMapControlsService.getZoomPercentage(distance)

            let actualZoomPercentage: number
            threeMapControlsService.zoomPercentage$.pipe(take(1)).subscribe(zoomPercentage => (actualZoomPercentage = zoomPercentage))

            expect(actualZoomPercentage).toBeCloseTo(expectedZoomPercentage)
        })
    })

    describe("getZoomPercentage", () => {
        beforeEach(() => {
            threeMapControlsService.controls.minDistance = 100
            threeMapControlsService.controls.maxDistance = 1000
            threeMapControlsService.MAX_ZOOM = 200
            threeMapControlsService.MIN_ZOOM = 10
        })

        it("should return MAX_ZOOM if distance is less than or equal to minDistance", () => {
            expect(threeMapControlsService.getZoomPercentage(50)).toBe(200)
            expect(threeMapControlsService.getZoomPercentage(100)).toBe(200)
        })

        it("should return MIN_ZOOM if distance is greater than or equal to maxDistance", () => {
            expect(threeMapControlsService.getZoomPercentage(1000)).toBe(10)
            expect(threeMapControlsService.getZoomPercentage(1500)).toBe(10)
        })

        it("should return a zoom percentage between MIN_ZOOM and MAX_ZOOM for a distance within the min and max range", () => {
            const result = threeMapControlsService.getZoomPercentage(550)

            expect(result).toBeGreaterThan(10)
            expect(result).toBeLessThan(200)
        })
    })

    describe("focusNode", () => {
        beforeEach(() => {
            withMockedThreeSceneService()
            withMockedControlService()
        })

        it("should save the current camera position and focus on the specified node", async () => {
            const nodePath = "some/node/path"
            const mockNodeBoundingSphere = new Sphere(new Vector3(10, 10, 10), 8.660_254_037_844_387)
            const mockNodeBoundingBox = new Box3().setFromCenterAndSize(new Vector3(10, 10, 10), new Vector3(10, 10, 10))

            const mockNode = {
                boundingBox: mockNodeBoundingBox,
                boundingSphere: {
                    getBoundingSphere: jest.fn().mockReturnValue(mockNodeBoundingSphere)
                }
            }

            const mockMapMesh = {
                getBuildingByPath: jest.fn().mockReturnValue(mockNode)
            }

            jest.spyOn(threeSceneService, "getMapMesh").mockReturnValue(mockMapMesh as unknown as CodeMapMesh)

            const spyEnsureProperDistanceAndFocus = jest.spyOn(threeMapControlsService as never, "ensureProperDistanceAndFocus")

            threeMapControlsService.focusNode(nodePath)

            expect(threeMapControlsService.positionBeforeFocus).toEqual(threeCameraService.camera.position.clone())
            expect(mockMapMesh.getBuildingByPath).toHaveBeenCalledWith(nodePath)
            await wait(0)
            expect(spyEnsureProperDistanceAndFocus).toHaveBeenCalledWith(mockNodeBoundingSphere)
        })
    })

    describe("unfocusNode", () => {
        beforeEach(() => {
            withMockedControlService()
        })

        it("should move the camera back to positionBeforeFocus if defined", async () => {
            const previousPosition = new Vector3(0, 0, 100)
            threeMapControlsService.positionBeforeFocus = previousPosition.clone()
            threeCameraService.camera.position.set(10, 10, 10)

            threeMapControlsService.unfocusNode()
            await wait(1100)

            expect(threeCameraService.camera.position).toEqual(previousPosition)
        })
    })
})
