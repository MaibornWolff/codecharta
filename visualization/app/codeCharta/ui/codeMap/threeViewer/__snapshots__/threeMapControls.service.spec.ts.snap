// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeMapControlsService rotateCameraInVectorDirection  1`] = `
Vector3 {
  "x": 0,
  "y": -8.076685518445746,
  "z": 0.0008076685518441267,
}
`;
